File: Source/Aladdin.Math.Java/src/aladdin/math/Field.java
package aladdin.math;
import java.lang.reflect.*;
import java.math.*;

////////////////////////////////////////////////////////////////////////////
// Интерфейс поля
////////////////////////////////////////////////////////////////////////////
public abstract class Field<E> extends Ring<E> implements IField<E>
{
    ///////////////////////////////////////////////////////////////////////
    // Кодирование элементов
    ///////////////////////////////////////////////////////////////////////
    
    // преобразовать в число
    // public abstract BigInteger toBigInteger(E element); 
    // преобразовать из числа
    // public abstract E fromBigInteger(BigInteger element) throws IOException; 
    
    ///////////////////////////////////////////////////////////////////////
    // Операции поля
    ///////////////////////////////////////////////////////////////////////
    
    // вычислить обратный элемент
    @Override public abstract E invert(E a);
              
    // умножить на обратный элемент
    @Override public E divide(E a, E b)  
    {
        // умножить на обратный элемент
        return (isOne(b)) ? a : product(a, invert(b)); 
    }
    // вычисление степени элемента 
    @Override public E power(E a, BigInteger e)
    {
        // обработать отрицательный сомножитель
        if (e.signum() < 0) return invert(power(a, e.abs())); 
        
        // вызвать базовую функцию
        return super.power(a, e); 
    }
    // произведение степеней элементов 
    @Override public E power_product(E P, BigInteger a, E Q, BigInteger b)
    {
        // проверить корректность данных
        if (a.signum() < 0) { P = invert(P); a = a.abs(); }
        if (b.signum() < 0) { Q = invert(Q); b = b.abs(); }
        
        // вызвать базовую функцию
        return super.power_product(P, a, Q, b); 
    }
    ///////////////////////////////////////////////////////////////////////
    // Операции кольца
    ///////////////////////////////////////////////////////////////////////
    @Override public final E remainder (E a, E b) { return zero(); } 
    
    // вычисление частного и остатка
    @SuppressWarnings({"unchecked"}) 
    @Override public final E[] divideAndRemainder(E a, E b)
    {
		// создать результирующий массив
        E[] result = (E[])Array.newInstance(a.getClass(), 2); 

        // вычисление частного и остатка
        result[0] = divide(a, b); result[1] = zero(); return result; 
    }
}

File: Source/Aladdin.Math/Math/Field.cs
using System; 

namespace Aladdin.Math 
{
    ////////////////////////////////////////////////////////////////////////////
    // Интерфейс поля
    ////////////////////////////////////////////////////////////////////////////
    [Serializable]
    public abstract class Field<E> : Ring<E>, IField<E>
    {
        // размерность поля
        public abstract int Dimension { get; }

        // вычислить обратный элемент
        public abstract E Invert(E a);
              
        // умножить на обратный элемент
        public override E Divide(E a, E b)  
        {
            // умножить на обратный элемент
            return (IsOne(b)) ? a : Product(a, Invert(b)); 
        }
        // вычисление степени элемента 
        public override E Power(E a, BigInteger e)
        {
            // обработать отрицательный сомножитель
            if (e.Signum < 0) return Invert(Power(a, e.Abs())); 
        
            // вызвать базовую функцию
            return base.Power(a, e); 
        }
        // произведение степеней элементов 
        public override E PowerProduct(E P, BigInteger a, E Q, BigInteger b)
        {
            // проверить корректность данных
            if (a.Signum < 0) { P = Invert(P); a = a.Abs(); }
            if (b.Signum < 0) { Q = Invert(Q); b = b.Abs(); }
        
            // вызвать базовую функцию
            return base.PowerProduct(P, a, Q, b); 
        }
        ///////////////////////////////////////////////////////////////////////
        // Операции кольца
        ///////////////////////////////////////////////////////////////////////
        public override E Remainder (E a, E b) { return Zero; } 
    
        // вычисление частного и остатка
        public override E[] DivideAndRemainder(E a, E b)
        {
            // вычисление частного и остатка
            return new E[2] { Divide(a, b), Zero }; 
        }
    }
}